### PR TITLE
2024 06 14 backtesting

### PR DESCRIPTION
**Pull request by hardyjosh**
## Motivation

Fix for https://github.com/rainlanguage/rain.orderbook/issues/674

## Solution

Implemented simple "backtesting", ie being able to run simulations across a range of blocks. Most of the work was create the deserializer for the syntax we wanted to use, and converting the range syntax into a concrete `Vec` of `BlockNumbers`. 

After that adding to the actual simulation loop was straightforward, except that Foundry's `roll_fork` doesn't actually update the env blocknumber so the rainlang word `block-number()` wasn't working.

To fix that I had to update the env block number every time we roll here: https://github.com/rainlanguage/rain.interpreter/pull/251

That PR needs to be merged first before merging this PR. 

## Checks

By submitting this for review, I'm confirming I've done the following:
- [x] made this PR as small as possible
- [x] unit-tested any new functionality
- [x] linked any relevant issues or PRs

### CODE REVIEW

File: crates/common/src/fuzz/mod.rs [hunk 1]
```diff
@@ -211,55 +229,59 @@ impl FuzzRunner {
...
+                let scenario_bindings = scenario_bindings.clone();
+                let dotrain = Arc::clone(&dotrain);
+
+                let mut final_bindings: Vec<Rebind> = vec![];
+
+                // For each scenario.fuzz_binds, add a random value
+                for elided_binding in elided_binding_keys.as_slice() {
+                    let mut val: [u8; 32] = [0; 32];
+                    self.rng.fill_bytes(&mut val);
+                    let hex = format!("0x{}", alloy_primitives::hex::encode(val));
```
Comment (thedavidmeister) [inline crates/common/src/fuzz/mod.rs:252]: does this ensure even number of hex nibbles on the output? i.e. left pad with 0 til even number of chars

File: crates/settings/src/blocks.rs [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+}
+
+impl Default for Blocks {
+    fn default() -> Self {
+        Blocks::RangeWithInterval {
+            range: BlockRange {
+                start: Block::Genesis,
+                end: Block::Latest,
+            },
+            interval: 1,
```
Comment (thedavidmeister) [inline crates/settings/src/blocks.rs:176]: hmm, a safer default would be start/end both at latest, this will expand out to millions of iterations probably

**Review by thedavidmeister — CHANGES_REQUESTED** [review]

File: crates/settings/src/blocks.rs (hunk 2 above)
Comment (hardyjosh) [reply crates/settings/src/blocks.rs:176]: ah yeah fair, i put this in for other tests where we need a default, but actually it's an `Option` in Scenario anyway, so I can delete this.

File: crates/common/src/fuzz/mod.rs (hunk 1 above)
Comment (hardyjosh) [reply crates/common/src/fuzz/mod.rs:252]: i've added padding to the format string

**Comment by hardyjosh** [comment]
@thedavidmeister assigning this to you so you can resolve the contract import issues

**Pull request merged**